File: src/sim/canalRuntimeReplan.ts
import { GAME_DAYS_PER_WEEK, REAL_MS_PER_GAME_WEEK } from "./canonicalTime";
import { allocateCanalBlockEconomy } from "./canalBlockEconomy";
import { calculateCanalBlockWear } from "./canalBlockWear";
import { buildCanalOperatingBlocks } from "./canalOperatingBlocks";
import type { GameState } from "./game";
import { planCanalOperations } from "./canalOperatingPlan";
import { operatingBlockKey, type OperatingWeekRuntime, type RuntimeOperatingBlock } from "./canalWeekRuntime";

function blockSettlesAt(startedAt: number, week: number, block: RuntimeOperatingBlock) {
  const weekStart = startedAt + (week - 1) * REAL_MS_PER_GAME_WEEK;
  const realMsPerGameDay = REAL_MS_PER_GAME_WEEK / GAME_DAYS_PER_WEEK;
  const gameHoursFromWeekStart = block.dayIndex * 24 + block.endsAt;
  return weekStart + (gameHoursFromWeekStart / 24) * realMsPerGameDay;
}

/** Rebuild only still-future blocks after a canonical command/world change. */
export function replanFutureCanalOperatingRuntime(
  snapshot: GameState,
  runtime: OperatingWeekRuntime,
  startedAt: number,
  at: number,
): OperatingWeekRuntime {
  if (runtime.week !== snapshot.week) return runtime;

  const settled = runtime.plannedBlocks.filter((block) => runtime.settledBlockKeys.includes(operatingBlockKey(block)));
  const freshPlan = planCanalOperations(snapshot);
  const freshAll = allocateCanalBlockEconomy(
    buildCanalOperatingBlocks(snapshot, freshPlan),
  ).map((block) => ({
    ...block,
    wear: calculateCanalBlockWear(snapshot, block),
  })) as RuntimeOperatingBlock[];
  const future = freshAll.filter(
    (block) => blockSettlesAt(startedAt, runtime.week, block) > at && !runtime.settledBlockKeys.includes(operatingBlockKey(block)),
  );

  return {
    ...runtime,
    plannedBlocks: [...settled, ...future].sort((a, b) =>
      blockSettlesAt(startedAt, runtime.week, a) - blockSettlesAt(startedAt, runtime.week, b),
    ),
  };
}

File: native/SaunaCore/Sources/SaunaCore/CanalRuntimeReplan.swift
import Foundation

public enum CanalRuntimeReplan {
    public static func blockSettlesAt(
        startedAt: Int64,
        week: Int,
        block: OperatingBlock
    ) -> Int64 {
        CanonicalTime.timestamp(
            startedAt: startedAt,
            gameWeek: week,
            dayIndex: block.dayIndex,
            gameHour: block.endsAt
        )
    }

    /// Rebuild only unsettled future blocks after the caller has already advanced canonical time
    /// to `at`. Settled keys and accrued accounting remain immutable history.
    public static func replanFuture(
        input: CanalOperatingInput,
        runtime: OperatingWeekRuntime,
        startedAt: Int64,
        at: Int64
    ) -> OperatingWeekRuntime {
        let settled = runtime.plannedBlocks.filter { runtime.settledBlockKeys.contains($0.key) }
        let fresh = CanalBlockPlanner.plan(input).filter {
            blockSettlesAt(startedAt: startedAt, week: runtime.week, block: $0) > at
                && !runtime.settledBlockKeys.contains($0.key)
        }
        var next = runtime
        next.plannedBlocks = (settled + fresh).sorted {
            blockSettlesAt(startedAt: startedAt, week: runtime.week, block: $0)
                < blockSettlesAt(startedAt: startedAt, week: runtime.week, block: $1)
        }
        return next
    }
}
